storage/Devices/PartitionImpl.cc

```diff
@@ -133,11 +133,29 @@ namespace storage
     unsigned int
     Partition::Impl::get_number() const
     {
-	string::size_type pos = get_name().find_last_not_of("0123456789");
-	if (pos == string::npos || pos == get_name().size() - 1)
+	const string& name = get_name();
+
+	string::size_type pos = name.find_last_not_of("0123456789");
+	if (pos == string::npos || pos == name.size() - 1)
 	    ST_THROW(Exception("partition name has no number"));
 
-	return atoi(get_name().substr(pos + 1).c_str());
+	return atoi(name.substr(pos + 1).c_str());
+    }
+
+
+    void
+    Partition::Impl::set_number(unsigned int number)
+    {
+	const string& name = get_name();
+
+	string::size_type pos = name.find_last_not_of("0123456789");
```

Comment:
IMHO that parsing from this and the previous function should be moved out into a separate function that can be tested separately. This looks much too fragile to me to be tested only inside such a relatively complex and specialized environment.

I propose a function to split a partition device name ("/dev/sda4" etc.) into the name part and the number part, i.e. "/dev/sda" and 4 (int). It whould throw exceptions upon a parse error. It might return a `std::pair<string, int>` or use non-const references as parameters. It would be a static (class) function. It might even be useful to the outside world, so it might be part of some StringUtil class.

Then it would be trivial to add sufficient unit tests to make sure it really works under all circumstances. I really don't like sequences like that `find_last_not_of("0123456789")` being repeated all over the code.